### DISCUSSION TITLE
How to prevent strays/unconnected records with many-to-many relations?

### DISCUSSION THREAD

**Discussion by justin-harr** [prisma/prisma#20831] [Q&A] [answered]
### Question

In my application my users need to be able to create posts with tags, add/remove tags to existing posts, and fetch all tags so that they can be displayed in a top-level filter and selected by the user to filter posts by tag. Following [the docs here](https://www.prisma.io/docs/guides/other/troubleshooting-orm/help-articles/working-with-many-to-many-relations), I created these models
```prisma
model User {
  id    Int    @id @default(autoincrement())
  name  String
  posts Post[]
}

model Post {
  id     Int    @id @default(autoincrement())
  title  String
  tags   Tag[]
  user   User?  @relation(fields: [userId], references: [id])
  userId Int?

  @@index([userId])
}

model Tag {
  id    Int    @id @default(autoincrement())
  name  String
  posts Post[]
}

```

Here is my create queries:
```ts
  await prisma.user.create({
    data: {
      name: "Bob",
    },
  });

  await prisma.post.create({
    data: {
      user: {
        connect: {
          id: 1,
        },
      },
      title: "How to become a butterfly",
      tags: {
        create: [{ name: "Magic" }, { name: "Butterflies" }],
      },
    },
  });
```

Then when someone updates the post I can do so like this:
```ts
  await prisma.post.update({
    where: {
      id: 1,
      user: {
        id: 1,
      },
    },
    data: {
      tags: {
        set: [{ id: 1 }, { id: 2 }],
        create: { name: "typescript" },
      },
    },
  });
```

So far so good but then lets say the user updates the post again by removing all the tags except one new tag, like so:
```ts
  await prisma.post.update({
    where: {
      id: 1,
      user: {
        id: 1,
      },
    },
    data: {
      tags: {
        set: [{ id: 1 }, { id: 2 }],
        create: { name: "awesome" },
      },
    },
  });
```

Now the three tags `dev`, `prisma`, `typescript` are completely disconnected from any posts and taking up database space without being used since when I use my find all tags query they are never returned:
```ts
await prisma.user.findMany({
    where: {
      id: 1,
    },
    include: {
      posts: {
        include: {
          tags: true,
        },
      },
    },
  });
```
Is there a way to make sure these extra records get cleaned up aside from running some recurring job to remove `tags` with no `post` relation?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @dasveloper 👋 

Thank you for raising this question.

Prisma does not automatically delete orphaned records (in this case, tags that are not associated with any posts). You would need to manually delete these records. This could be done by creating a function that finds all tags where the posts field is empty and deletes them. 

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
